core/raftlease/store_test.go

```diff
@@ -0,0 +1,469 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package raftlease_test
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/juju/pubsub"
+	"github.com/juju/testing"
+	jc "github.com/juju/testing/checkers"
+	gc "gopkg.in/check.v1"
+
+	"github.com/juju/juju/core/globalclock"
+	"github.com/juju/juju/core/lease"
+	"github.com/juju/juju/core/raftlease"
+	coretesting "github.com/juju/juju/testing"
+)
+
+type storeSuite struct {
+	testing.IsolationSuite
+
+	clock  *testing.Clock
+	target *fakeTarget
+	fsm    *fakeFSM
+	hub    *pubsub.StructuredHub
+	store  *raftlease.Store
+}
+
+var _ = gc.Suite(&storeSuite{})
+
+func (s *storeSuite) SetUpTest(c *gc.C) {
+	s.IsolationSuite.SetUpTest(c)
+
+	startTime, err := time.Parse(time.RFC3339, "2018-08-08T08:08:08+08:00")
+	c.Assert(err, jc.ErrorIsNil)
+	s.clock = testing.NewClock(startTime)
+	s.target = &fakeTarget{}
+	s.fsm = &fakeFSM{
+		leases:     make(map[lease.Key]lease.Info),
+		globalTime: s.clock.Now(),
+	}
+	s.hub = pubsub.NewStructuredHub(nil)
+	s.store = raftlease.NewStore(raftlease.StoreConfig{
+		FSM:          s.fsm,
+		Hub:          s.hub,
+		Target:       s.target,
+		RequestTopic: "lease.request",
+		ResponseTopic: func(reqID uint64) string {
+			return fmt.Sprintf("lease.request.%d", reqID)
+		},
+		Clock:          s.clock,
+		ForwardTimeout: time.Second,
+	})
+}
+
+func (s *storeSuite) TestClaim(c *gc.C) {
+	s.handleHubRequest(c,
+		func() {
+			err := s.store.ClaimLease(
+				lease.Key{"warframe", "rhino", "prime"},
+				lease.Request{"lotus", time.Second},
+			)
+			c.Assert(err, jc.ErrorIsNil)
+		},
+
+		raftlease.Command{
+			Version:   1,
+			Operation: raftlease.OperationClaim,
+			Namespace: "warframe",
+			ModelUUID: "rhino",
+			Lease:     "prime",
+			Holder:    "lotus",
+			Duration:  time.Second,
+		},
+		func(req raftlease.ForwardRequest) {
+			_, err := s.hub.Publish(
+				req.ResponseTopic,
+				raftlease.ForwardResponse{},
+			)
+			c.Check(err, jc.ErrorIsNil)
+		},
+	)
+	s.target.CheckCall(c, 0, "Claimed",
+		lease.Key{"warframe", "rhino", "prime"},
+		"lotus",
+	)
+}
+
+func (s *storeSuite) TestClaimTimeout(c *gc.C) {
+	s.handleHubRequest(c,
+		func() {
+			errChan := make(chan error)
+			go func() {
+				errChan <- s.store.ClaimLease(
+					lease.Key{"warframe", "vauban", "prime"},
+					lease.Request{"vor", time.Second},
+				)
+			}()
+			// Jump time forward further than the 1-second forward
+			// timeout.
+			s.clock.WaitAdvance(2*time.Second, coretesting.LongWait, 1)
+			select {
+			case err := <-errChan:
+				c.Assert(err, jc.Satisfies, lease.IsTimeout)
```

Comment:
We need a bool to prove we got here
(similar comment for other tests)

Reply:
I don't understand this - doesn't the channel write show that we got here? This select can only succeed if we get a timeout error on the channel, otherwise it will fatal, right?